torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp

```diff
@@ -398,12 +398,38 @@ std::shared_ptr<ReductionParams> innerPersistentHeuristic(
       : bdimx + (device_warp_size - bdimx % device_warp_size);
 
   bool pad_bdimx = bdimx > 16 &&
-      padded_bdimx * bdimy * bdimz <
-          (int64_t)at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock;
+      padded_bdimx * bdimy * bdimz < (int64_t)dev_prop->maxThreadsPerBlock;
 
   pad_bdimx = pad_bdimx &&
       bdimx * inner_reduction_unroll_factor != inner_most_dimension_numel;
 
+  // when register usage is high, set maxregcount to avoid low occupancy raito,
+  // will cause regisgter spills
+  int64_t nvrtc_register_per_thread = LaunchParams::UNINITIALIZED_VAL;
+  const int blocksPerKernel = godim;
+  const int persistent_buffer_size = batches_per_block_inner_reduction *
+      inner_reduction_unroll_factor * max_input_dtype_size;
+  const bool large_l2 =
+      dev_prop->l2CacheSize > 10 * 1024 * 1024; // filter out v100
+  if (vectorize && large_l2 && blocksPerKernel > device_multiprocessor_count &&
+      persistent_buffer_size >= 32) {
+    constexpr int reg_allocation_granularity = 256;
+    constexpr float occupancy_ratio = 0.4;
```

Comment:
Is this saying we are trying to achieve at least occupancy ratio of 0.4?

Reply:
yes. also tested 0.5, not much difference.

Reply:
Shouldn't the occupancy ratio also be affected by the persistent buffer size? It seems strange to me that we always set the occupancy target the same number no matter how large the persistent buffer size would be.

Reply:
I tested a lower occupancy ratio of 0.25 if buffer size is >=128. This can fully avoid register spills but lead to performance drop compared with occupancy ratio of 0.40 which has register spills.
![image](https://user-images.githubusercontent.com/116412316/205510809-131c1504-9e8a-443c-9649-4af45a483a4c.png)

Reply:
I wonder if this is something then we should go and try to retune the heuristics for. Could be a non-linear effect.